ArmVirtPkg/Library/ArmPlatformDeviceInfoLib/ArmPlatformDeviceInfoLib.inf

```diff
@@ -28,6 +28,8 @@
   HwInfoParser/FdtInfoParser.h
   HwInfoParser/FdtUtility.c
   HwInfoParser/FdtUtility.h
+  HwInfoParser/Pci/PciConfigSpaceParser.c
+  HwInfoParser/Pci/PciConfigSpaceParser.h

```

Comment:
Why is this in the Arm CCA PR?  On a quick glance this looks like some independent change.

Also why put this into ArmVirt?  We have other platforms like riscv which use device trees too and can probably use this code too, so I think OvmfPkg would be a better place for it.

Reply:
Hi Gerd,

Thank you for reviewing the patch series and for the feedback.

> Why is this in the Arm CCA PR? On a quick glance this looks like some independent change.

Yes, I agree this is an independent change and is a general improvement for normal VMs as well. However, this was introduced as this a requirement for a Realm VM, please see 
https://github.com/tianocore/edk2/pull/6480/commits/43476c3d268b360e72e4aa55aa47e8f28a87da87 and therefore submitted along with this series.
However, please do let me know if you think I should submit this as a separate patch.

> Also why put this into ArmVirt? We have other platforms like riscv which use device trees too and can probably use this code too, so I think OvmfPkg would be a better place for it.

I agree this can be utilised by other architectures as well and it can be moved to OvmfPkg. However, I think there may be some additional changes required for supporting other architectures, e.g. there is a pending series that should help pave the way for this change, see https://github.com/tianocore/edk2/pull/6053/commits/a1f26543bbd52286733501a28a95880d48ce5b50

Since this patch is a subset of the FdtHwInfoParser, I think similar changes would apply here as well. Considering that, I think it would be good to pursue this effort once the RiscV support is enabled in Dynamic Tables. 

Please do let me know if you think otherwise, and I will update this PR accordingly.

Regards,

Sami Mujawar

Reply:
Phew.  Is there some overview how the hardware detection works in for Realm VMs?  Apparently this is different from the usual qemu workflow, where qemu generates both FDT and ACPI tables, the guest firmware simply downloads them via FwCfg and installs them in guest memory.

So Realm VMs get a FDT only apparently (where does it come from?), then use DynamicTablesPkg to generate the ACPI tables inside the guest, correct?  Where exactly in this workflow is this change needed?  Parse the FDT and generate descriptions tables which DynamicTablesPkg can consume?

On riscv:  Yes, sure, additinal changes will be needed to have the riscv firmware actually use this (and this is obviously beyond the scope of this PR)..  But if this is something which is likely to happen we should merge the code into OvmfPkg right away instead of merging to ArmVirt and move over to OvmfPkg later on.

Reply:
I believe the ACPI table generation is specific to the ArmVirtKvmTool. If I remember correctly, ArmVirtQemu and ArmVirtCloudHv don't have it, so it's not (yet) a property of Realm VMs, only Kvmtool VMs. At the moment in my experimental Realm support for ArmVirtQemu, QEMU still generates ACPI tables in addition to FDT (though I haven't tested ACPI boot).

There is a larger question of whether we should create a single edk2 image for all Realm VMs, based on ArmVirtKvmtool but supporting QEMU, cloud-hypervisor and others. Given that Realm VMs have specific needs (initialization through RMM RSI, early attestation, lack of NV storage, reduced attack surface...) it would be nice to have all those things in a single image rather than porting them to each existing ArmVirt*. I've been meaning to investigate this but haven't found the time.

Having FDT->ACPI generation on this single image could also be useful for attestation. Given that firmware tables are measured (to ensure the untrusted host does not include harmful data in them), a verifier needs to be able to independently reconstruct the firmware tables, and it would reduce the efforts needed by verifiers to only measure one set of firmware tables rather than both FDT and ACPI. In QEMU the FDT is placed in memory, and ACPI tables are sent via FwCfg (whose address is found in the FDT). I believe CloudHV places both in memory and the project doesn't want to implement FwCfg.

This would require FDT to be able to represent everything ACPI can (at least for arm64 VMs), and some things may be missing. vCPU hotplug is a recent feature on Arm and I believe still requires ACPI at the moment. There might be others.

Reply:
One of the reasons qemu generates ACPI tables for the guest is to decouple firmware and qemu, i.e. qemu can add new features which require ACPI table updates without depending on firmware updates.

This might be less of a concern for Realm VMs.  I'm wondering whenever it is actually possible to support cpu hotplug (or memory hotplug) for Realm VMs, and there are probably more features which are not useful for CC guests.

So the idea to have a single Realm VM firmware image and have the image generate ACPI tables from FDT looks workable to me.

Reply:
While being at it:  Setting `PcdConfidentialComputingGuestAttr` probably makes sense.